### PR TITLE
[multibyte] Add multibyte array store instructions.

### PR DESCRIPTION
**Pull request by brendandahl**
Prototype implementation of the new multibyte array proposal that reuses the existing memory instructions.
The syntax for the new instructions is as follows:
  `<i32|i64|f32|f64>.store (type $type_idx) <array ref> <index> <value>`

Some spec related TODOs:
  - Decide on new instructions vs existing memory instructions with memarg flag.
  - If using existing instructions:
      - What does the wat syntax look like? e.g. `i32.store type=array x y` or `i32.store array x y` or ???
      - Do we support immediate offsets? - Do we support i32.store8? (effectively the same as array.set on i8)
  - Do we support atomic instructions?

Some implementation TODOs:
 - Add feature flag. 
 - Add v128.store.
  - Go through the various `visitArrayStore` functions and implement them or make sure the copied versions makes sense.
  - Add fuzzing support.
  - Make it so there's only one `WasmBinaryReader::getMemarg`
  - Add JS API

### CODE REVIEW

File: src/wasm.h [hunk 1]
```diff
@@ -1827,6 +1833,22 @@ class ArraySet : public SpecificExpression<Expression::ArraySetId> {
...
+  ArrayStore() = default;
+  ArrayStore(MixedArena& allocator) {}
+
+  uint8_t bytes;
+  Expression* ref;
+  Expression* index;
+  Expression* value;
+  Type valueType;
+  // TODO is this needed?
+  MemoryOrder order = MemoryOrder::Unordered;
```
Comment (tlively) [inline src/wasm.h]: Probably in the long run, but not urgently. I would leave it out for now.

File: test/lit/array-multibyte.wast [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+  ;; ROUNDTRIP:      (global $arr (ref $i8_array) (array.new_default $i8_array
+  ;; ROUNDTRIP-NEXT:  (i32.const 4)
+  ;; ROUNDTRIP-NEXT: ))
+  (global $arr (ref $i8_array)
+    (array.new_default $i8_array (i32.const 4))
+  )
+
+  ;; CHECK:      (func $stores (type $1)
+  ;; CHECK-NEXT:  (i32.store8 type=array
```
Comment (tlively) [inline test/lit/array-multibyte.wast]: The text format is going to need a type index immediate to give the type of the array that is being accessed. Maybe we can just use the presence of that immediate to signal that we're accessing an array? So we might have `i32.store8 $myArray` instead of `i32.store8 type=array`. Obviously we can't settle that here, but we're definitely going to need the immediate.

File: src/wasm/wasm-binary.cpp [hunk 3]
```diff
@@ -3560,40 +3560,40 @@ Result<> WasmBinaryReader::readInst() {
       return builder.makeLoad(8, false, offset, align, Type::f64, mem);
     }
     case BinaryConsts::I32StoreMem8: {
-      auto [mem, align, offset] = getMemarg();
-      return builder.makeStore(1, offset, align, Type::i32, mem);
+      auto [mem, align, offset, backing] = getMemargWithBacking();
```
Comment (tlively) [inline src/wasm/wasm-binary.cpp]: The binary format will need the type immediate as well, so this will have to look something like this:

```suggestion
      auto [mem, align, offset, backing] = getMemargWithBacking();
      if (backing == BackingType::Array) {
        HeapType type = getIndexedHeapType();
        ...
```

At that point we might as well add a separate `IRBuilder` method for making array stores, which should help separate concerns more.

File: test/lit/array-multibyte.wast (hunk 2 above)
Comment (brendandahl) [reply test/lit/array-multibyte.wast]: What's the reason for the type immediate if it's only ever going to be an `(array (mut i8))`? I guess it could be a memref in the future or?

Comment (tlively) [reply test/lit/array-multibyte.wast]: There is an infinite set of `(array (mut i8))` types, and I would expect these instructions to work with all of them. For example, `$A`, `$B`, `$C`, and `$D` are all different types here:

```
(type $A (sub final (array (mut i8))))
(type $B (sub (array (mut i8))))
(type $C (sub $B (array (mut i8))))
(rec (type $D (array (mut i8))) (type (struct)))
```

Given that there are several types that could be accepted, we need the type index to make sure we can look at an instruction with its immediates in isolation and say what its input and output types will be. (This ensures we maintain the [principal types](https://webassembly.github.io/spec/core/appendix/properties.html#principal-types) property, which is important for analyzing instruction sequences in isolation. For example, it is important for outlining.)

Comment (tlively) [reply test/lit/array-multibyte.wast]: The alternative is to accept just a single, well-known array type, e.g. `(rec (type (sub final (array (mut i8))))).0`, but that would be somewhat limiting for producers, would require extra work in Binaryen, and would be unprecedented for core Wasm instructions.

Comment (brendandahl) [reply test/lit/array-multibyte.wast]: I started playing with `i32.store8 $myArray`, but if I'm following this correctly, I don't think there's a way to distinguish multimemory from array store e.g. `i32.store8 1` could be a memory or type index.

Comment (tlively) [reply test/lit/array-multibyte.wast]: oh yeah, good point :/ I guess we do need to distinguish this somehow. For prototyping I guess we can do whatever, but it would also be good to open a issue on the proposal repo to see what folks think is best here.

Comment (brendandahl) [reply test/lit/array-multibyte.wast]: Coming back to this... another option is we could require parentheses around the type. e.g. `i32.load ($type_index) (<array ref>) (<address>)`. Then we could differentiate between memidx. This came to mind after you mentioned not enough parentheses :).

Anyway none of this is critical to the prototype, but thought that may be an interesting option.

Comment (tlively) [reply test/lit/array-multibyte.wast]: I agree we should do whatever we want for now in Binaryen to avoid blocking on text format issues. So far the text format always has some keyword token at the beginning of parenthetical phrases, though, so I don't think that particular option is the best fit for the final choice.

File: src/parser/parsers.h [hunk 4]
```diff
@@ -1786,6 +1786,18 @@ Result<> makeStore(Ctx& ctx,
...
+    std::optional<typename Ctx::HeapTypeT> arrayType;
+    auto x = typeidx(ctx);
+    CHECK_ERR(x);
+
+    if (!ctx.in.takeRParen()) {
+      return ctx.in.err("expected end of type use");
+    }
+
+    arrayType = *x;
+    return ctx.makeArrayStore(pos, annotations, type, bytes, *arrayType);
```
Comment (tlively) [inline src/parser/parsers.h:1829]:
```suggestion
    return ctx.makeArrayStore(pos, annotations, type, bytes, *x);
```
It doesn't look like `arrayType` should be necessary here.

File: src/parser/parsers.h [hunk 5]
```diff
@@ -1786,6 +1786,18 @@ Result<> makeStore(Ctx& ctx,
                    Type type,
                    int bytes,
                    bool isAtomic) {
+  if (ctx.in.takeSExprStart("type"sv)) {
+    std::optional<typename Ctx::HeapTypeT> arrayType;
+    auto x = typeidx(ctx);
+    CHECK_ERR(x);
```
Comment (tlively) [inline src/parser/parsers.h]: Maybe rename `x` to `arrayType`?

File: src/ir/subtype-exprs.h [hunk 6]
```diff
@@ -408,6 +408,13 @@ struct SubtypingDiscoverer : public OverriddenVisitor<SubType> {
     auto array = curr->ref->type.getHeapType().getArray();
     self()->noteSubtype(curr->value, array.element.type);
   }
+  void visitArrayStore(ArrayStore* curr) {
+    if (!curr->ref->type.isArray()) {
+      return;
+    }
+    auto array = curr->ref->type.getHeapType().getArray();
+    self()->noteSubtype(curr->value, array.element.type);
+  }
```
Comment (tlively) [inline src/ir/subtype-exprs.h]: Assuming this is only valid for numeric types, there are no interesting subtype relationships possible here, so we can just do nothing here like we do for `Store`.

File: src/wasm.h [hunk 7]
```diff
@@ -1860,6 +1866,20 @@ class ArraySet : public SpecificExpression<Expression::ArraySetId> {
...
+class ArrayStore : public SpecificExpression<Expression::ArrayStoreId> {
+public:
+  ArrayStore() = default;
+  ArrayStore(MixedArena& allocator) {}
+
+  uint8_t bytes;
+  Expression* ref;
+  Expression* index;
+  Expression* value;
+  Type valueType;
```
Comment (tlively) [inline src/wasm.h]: I see that this field is also used in `Store`. Does it add any information that can't be recovered from `value->type`, though? (Besides in the case where `value->type` is unreachable.)

File: src/wasm/wasm-stack.cpp [hunk 8]
```diff
@@ -56,6 +56,70 @@ void BinaryInstWriter::emitIfElse(If* curr) {
   o << static_cast<int8_t>(BinaryConsts::Else);
 }
 
+void BinaryInstWriter::emitStore(uint8_t bytes, Type valueType) {
```
Comment (tlively) [inline src/wasm/wasm-stack.cpp]:
```suggestion
void BinaryInstWriter::emitStoreOpcode(uint8_t bytes, Type valueType) {
```

File: src/wasm/wasm-ir-builder.cpp [hunk 9]
```diff
@@ -2367,6 +2374,16 @@ Result<> IRBuilder::makeArraySet(HeapType type, MemoryOrder order) {
   return Ok{};
 }
 
+Result<>
+IRBuilder::makeArrayStore(HeapType arrayType, unsigned bytes, Type type) {
+  ArrayStore curr;
+  curr.valueType = type;
+  CHECK_ERR(
+    ChildPopper{*this}.visitArrayStore(&curr, HeapTypes::getMutI8Array()));
```
Comment (tlively) [inline src/wasm/wasm-ir-builder.cpp]: We can use the real type here, but it might also require adding validatoin that the real type is an array. See e.g. `makeArrayGet` and `makeArraySet`.
```suggestion
    ChildPopper{*this}.visitArrayStore(&curr, arrayType));
```

File: test/lit/array-multibyte.wast [hunk 10]
```diff
@@ -0,0 +1,126 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+
+;; Check that array types and operations are emitted properly in the binary format.
+
+;; RUN: wasm-opt %s -all -S -o - | filecheck %s
+;; RUN: wasm-opt %s -all --roundtrip -S -o - | filecheck %s --check-prefix=ROUNDTRIP
```
Comment (tlively) [inline test/lit/array-multibyte.wast]: Sometimes we use the `RTRIP` prefix because it has the same number of characters as `CHECK` and then everything lines up nicely.
```suggestion
;; RUN: wasm-opt %s -all --roundtrip -S -o - | filecheck %s --check-prefix=RTRIP
```

File: test/lit/array-multibyte.wast [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  (func $stores
+    (i32.store8 (type $i8_array) (global.get $arr) (i32.const 0) (i32.const 0))
+    (i32.store16 (type $i8_array) (global.get $arr) (i32.const 0) (i32.const 0))
+    (i32.store (type $i8_array) (global.get $arr) (i32.const 0) (i32.const 0))
+    (f32.store (type $i8_array) (global.get $arr) (i32.const 0) (f32.const 0))
+
+    (i64.store8 (type $i8_array) (global.get $arr) (i32.const 0) (i32.const 0))
+    (i64.store16 (type $i8_array) (global.get $arr) (i32.const 0) (i32.const 0))
+    (i64.store32 (type $i8_array) (global.get $arr) (i32.const 0) (i32.const 0))
+    (f64.store (type $i8_array) (global.get $arr) (i32.const 0) (f64.const 0))
```
Comment (tlively) [inline test/lit/array-multibyte.wast]: It would be good to use different constants for the two i32 operands to make sure they are parsed in the correct order.

It would also be good to check the cases where the ref operand is null or unreachable.

File: test/spec/array-multibyte.wast [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+  (func $f64_set (export "f64_set") (param $index i32) (param $value f64)
+    (f64.store (type $i8_array)
+      (global.get $arr_8)
+      (local.get $index)
+      (local.get $value)
+    )
+  )
+
+  ;; ??? Do we even want to spec out this instruction since array.store is the
+  ;; same thing?
```
Comment (tlively) [inline test/spec/array-multibyte.wast]:
```suggestion
  ;; ??? Do we even want to spec out this instruction since array.set is the
  ;; same thing?
```
Good question!

File: test/spec/array-multibyte.wast [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+      (local.get $value)
+    )
+    ;; TODO: when multibyte get is supported use that instead here.
+    (call $load_i64_from_array (global.get $arr_8) (i32.const 0))
+    (f64.reinterpret_i64)
+  )
+)
+
+;;
+;; 32 bit round trip tests
```
Comment (tlively) [inline test/spec/array-multibyte.wast:275]: In addition to the `set_and_get` calls, it would be good to test a few raw stores followed by normal `array.gets` to show that the individual elements have been updated correctly. It would also be good to test "unaligned" stores at non-zero offsets.

File: test/spec/array-multibyte.wast [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+(assert_return (invoke "i32_set_and_get_i32" (i32.const 256)) (i32.const 256))
+(assert_return (invoke "i32_set_and_get_i32" (i32.const -1)) (i32.const -1))
+(assert_return (invoke "i32_set_and_get_i32" (i32.const 2147483647)) (i32.const 2147483647))
+(assert_return (invoke "i32_set_and_get_i32" (i32.const -2147483648)) (i32.const -2147483648))
+
+(assert_return (invoke "set_and_get_f32" (f32.const 0)) (f32.const 0))
+(assert_return (invoke "set_and_get_f32" (f32.const -1)) (f32.const -1))
+(assert_return (invoke "set_and_get_f32" (f32.const 3.3)) (f32.const 3.3))
+(assert_return (invoke "set_and_get_f32" (f32.const -2.000000238418579)) (f32.const -2.000000238418579))
+(assert_return (invoke "set_and_get_f32" (f32.const nan)) (f32.const nan))
```
Comment (tlively) [inline test/spec/array-multibyte.wast:299]: You could also test nans with specific payloads.

File: test/spec/array-multibyte.wast [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+;; Valid range: [0, 3]
+(assert_trap (invoke "i32_set_i8" (i32.const -1) (i32.const 0)) "out of bounds")
+(assert_return (invoke "i32_set_i8" (i32.const 0) (i32.const 0)))
+(assert_return (invoke "i32_set_i8" (i32.const 1) (i32.const 0)))
+(assert_return (invoke "i32_set_i8" (i32.const 2) (i32.const 0)))
+(assert_return (invoke "i32_set_i8" (i32.const 3) (i32.const 0)))
+(assert_trap (invoke "i32_set_i8" (i32.const 4) (i32.const 0)) "out of bounds")
+
+;; i32_set_i16: Writes 2 bytes
+;; Valid range: offset + 2 <= 4 -> Max offset 2
```
Comment (tlively) [inline test/spec/array-multibyte.wast:399]: It would be good to check the (absence of) partial writes in these partial-OOB cases.

File: src/wasm.h (hunk 7 above)
Comment (brendandahl) [reply src/wasm.h]: No I don't think it does, it should be a mirror. In the unreachable case, we'd not know what to emit in `BinaryInstWriter::emitStoreOpcode` if the type was lost, but would we ever hit that case?

Comment (tlively) [reply src/wasm.h]: Let's try removing it and seeing what happens. If it works out, we can add a TODO to do the same thing for Store.

Comment (brendandahl) [reply src/wasm.h]: Removed in: https://github.com/WebAssembly/binaryen/pull/8059/commits/6f4b1c7fbcfb7bbf09ed3a9d2dee41f6deb3ab46

File: src/wasm/wasm-ir-builder.cpp [hunk 16]
```diff
@@ -567,10 +567,14 @@ struct IRBuilder::ChildPopper
...
-                           std::optional<HeapType> ht = std::nullopt) {
+  Result<>
+  visitArrayStoreWithType(ArrayStore* curr, HeapType arrayType, Type type) {
     std::vector<Child> children;
-    ConstraintCollector{builder, children}.visitArrayStore(curr, ht);
+    // Push in Wasm stack order (ref, index, value).
+    // popConstrainedChildren processes them in reverse order.
+    children.push_back({&curr->ref, {Type(arrayType, Nullable)}});
+    children.push_back({&curr->index, {Type::i32}});
+    children.push_back({&curr->value, {type}});
```
Comment (tlively) [inline src/wasm/wasm-ir-builder.cpp]: This should be handled by ChildTyper. Is it not?

File: test/lit/array-multibyte.wast [hunk 17]
```diff
@@ -3,124 +3,429 @@
...
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (i32.const 1)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (f64.const 2)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (unreachable)
+  ;; RTRIP-NEXT: )
+  (func $stores_null
+    (i32.store8 (type $i8_array) (ref.null $i8_array) (i32.const 1) (i32.const 2))
```
Comment (tlively) [inline test/lit/array-multibyte.wast:306]: It's unexpected that these get parsed as a drops + unreachable sequence. I would have expected them to be parsed ok, and then printed as blocks similar to these but with a comment about not being able to print the ArrayStore instruction.

File: test/lit/array-multibyte.wast [hunk 18]
```diff
@@ -3,124 +3,429 @@
...
+  ;; CHECK-NEXT:    (f64.const 2)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (unreachable)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $stores_unreachable (type $0)
+  ;; RTRIP-NEXT:  (unreachable)
+  ;; RTRIP-NEXT: )
+  (func $stores_unreachable
+    (i32.store8 (type $i8_array) (unreachable) (i32.const 1) (i32.const 2))
```
Comment (tlively) [inline test/lit/array-multibyte.wast:419]: Same here.

File: test/spec/array-multibyte.wast [hunk 19]
```diff
@@ -319,52 +330,106 @@
...
+(assert_return (invoke "set_and_get_f64" (f64.const -nan:0xedcba98765432)) (f64.const -nan:0xedcba98765432))
+
+;;
+;; Byte-wise store and unaligned store tests (32 bit)
+;;
+
+(invoke "i32_set_i32" (i32.const 0) (i32.const 0x00000000)) ;; clear
+(invoke "i32_set_i16" (i32.const 0) (i32.const 0x1234))
+(assert_return (invoke "get_array_4_byte" (i32.const 0)) (i32.const 0x34))
+(assert_return (invoke "get_array_4_byte" (i32.const 1)) (i32.const 0x12))
```
Comment (tlively) [inline test/spec/array-multibyte.wast:343]: Might as well check that bytes 2 and 3 are zeros, too.

File: src/wasm/wasm-ir-builder.cpp (hunk 16 above)
Comment (brendandahl) [reply src/wasm/wasm-ir-builder.cpp]: I may be missing something with how to make this work, but during makeArrayStore we can't access `curr->value` yet since it hasn't been popped yet and subsequently ChildTyper can't use it to determine pop constraints either. I guess this why the original `valueType` was also useful.

Keep in mind array.store can store various value types (i64, f32, f64, etc.) into an i8 array.

Comment (tlively) [reply src/wasm/wasm-ir-builder.cpp]: Ah, makes sense. Let's solve this by passing the expected `valueType` through as an additional optional parameter here, essentially treating it like an extra type immediate on the instruction.

File: src/wasm.h [hunk 20]
```diff
@@ -72,6 +72,11 @@ enum class MemoryOrder : uint8_t {
   AcqRel,
 };
 
+enum class BackingType {
+  Memory,
+  Array,
+};
```
Comment (kripken) [inline src/wasm.h]: Could this move out of `wasm.h`? It isn't used here, and we try to keep this header minimal

The users are wasm-stack/binary, so perhaps in one of those?

File: test/lit/array-multibyte.wast (hunk 17 above)
Comment (brendandahl) [reply test/lit/array-multibyte.wast:306]: Looks like it was being generated, but then removed by the last round trip text check since that loses the block comments. I've added another prefix.

Comment (brendandahl) [reply test/lit/array-multibyte.wast:306]: Alternatively, I could not do a filecheck on the last run, but just ensure it runs without an error I guess.

Comment (tlively) [reply test/lit/array-multibyte.wast:306]: The latter seems fine.

File: src/ir/child-typer.h [hunk 21]
```diff
@@ -1096,6 +1096,21 @@ template<typename Subtype> struct ChildTyper : OverriddenVisitor<Subtype> {
...
+    if (!ht) {
+      if (!curr->ref->type.isRef()) {
+        self().noteUnknown();
+        return;
+      }
+      ht = curr->ref->type.getHeapType();
+    }
+    note(&curr->ref, Type(*ht, Nullable));
+    note(&curr->index, Type::i32);
+    note(&curr->value, valueType ? *valueType : curr->value->type);
```
Comment (tlively) [inline src/ir/child-typer.h]: We should `self().noteUnknown()` if `curr->value->type == Type::unreachable`.

File: src/ir/possible-contents.cpp [hunk 22]
```diff
@@ -1102,6 +1102,13 @@ struct InfoCollector
     addChildParentLink(curr->ref, curr);
     addChildParentLink(curr->value, curr);
   }
+  void visitArrayStore(ArrayStore* curr) {
+    if (curr->ref->type == Type::unreachable) {
+      return;
+    }
+    addChildParentLink(curr->ref, curr);
+    addChildParentLink(curr->value, curr);
```
Comment (kripken) [inline src/ir/possible-contents.cpp:1110]: These links must be handled below, see new line 2837 for ArraySet.

File: src/passes/Print.cpp [hunk 23]
```diff
@@ -437,6 +437,25 @@ struct PrintExpressionContents
...
+          o << "16";
+        }
+      } else if (bytes == 4) {
+        o << "32";
+      } else {
+        abort();
+      }
+    }
+    return o;
+  }
```
Comment (kripken) [inline src/passes/Print.cpp:465]: Looks very similar to

https://github.com/brendandahl/binaryen/blob/324e152521fdf41bd09a5cf9b2bc1de151b8348a/src/passes/Print.cpp#L578-L590 

maybe these can be shared?

**Comment by brendandahl** [comment]
> Please fuzz before landing.

How long would you recommend letting it run for?

**Comment by kripken** [comment]
I would give it half a day, personally.

File: src/ir/possible-contents.cpp [hunk 24]
```diff
@@ -3236,6 +3247,12 @@ void Flower::writeToData(Expression* ref, Expression* value, Index fieldIndex) {
   // As in readFromData, normalize to the proper cone.
   auto cone = refContents.getCone();
   auto normalizedDepth = getNormalizedConeDepth(cone.type, cone.depth);
+  if (multibyte) {
```
Comment (brendandahl) [inline src/ir/possible-contents.cpp:3268]: @kripken Here's where I'm having issues. 
```wat
(module
 (type $0 (func))
 (type $1 (array (mut i8)))
 (global $global$0 (ref $1) (array.new_default $1
  (i32.const 4)
 ))
 (func $0
  (f64.store (type $1)
   (global.get $global$0)
   (i32.const 1)
   (f64.const 2)
  )
 )
)
```
with
`wasm-opt a.wasm -o b.wasm --discard-global-effects --roundtrip --signature-refining --gufa-cast-all --shrink-level=1 --closed-world --dce --closed-world -all`

Comment (kripken) [reply src/ir/possible-contents.cpp:3268]: I am guessing that `filterExpressionContents` applies the type of the expression in the wasm, and here it writes a different type, so the filtering ends up making things worse.

Updating `filterExpressionContents` should fix this. I assume the issue is on ArrayGet. When such a get sees data of another type in a numeric array, rather than combining them to get Many, we can apply `PossibleContents::fromType(value->type)`. That is, an unknown value in `array.get` is still going to be of the type that particular get reads.

Comment (brendandahl) [reply src/ir/possible-contents.cpp:3268]: Alright here's what I came up with in c148eaf03660944d0e8fb1f47c0b8000bd8bfdd3 . I'll let it cook on the fuzzer for a bit.

Comment (brendandahl) [reply src/ir/possible-contents.cpp:3268]: @kripken This has run against the fuzzer for a day. Did you want to review the above changes or is this good to land?

File: src/ir/possible-contents.cpp [hunk 25]
```diff
@@ -3017,6 +3022,11 @@ void Flower::filterDataContents(PossibleContents& contents,
     contents = PossibleContents::none();
     return;
   }
+  if (contents.isMany()) {
+    // An unknown state (e.g. from combining writes of different types like in
+    // multibyte array stores) translates into the most generic bounded type.
```
Comment (kripken) [inline src/ir/possible-contents.cpp:3030]: Please add a TODO like `TODO: We could optimize these, as e.g. a write of i16 0x1212 does not actually conflict with a write of i8 0x12`.

File: src/ir/possible-contents.cpp (hunk 24 above)
Comment (kripken) [reply src/ir/possible-contents.cpp:3268]: lgtm with the suggested TODO
